### PR TITLE
Feature: Being able to add support for 3D model

### PR DESCRIPTION
**Pull request by meiliang1123**
- Feature: Set support mesh as child of target mesh
- Feature: Calculate support height on moving
- Feature: Set support_xy_distance when modelGroup has support model
- Feature: Being able to move model on XY plane
- Feature: Add support continuously
- Improvement: Use intersect point as the avgZ of the triangle
- Feature: Use CuraEngine support_mesh feature to slice

### CODE REVIEW

File: src/app/flux/printing/DefinitionManager.js [hunk 1]
```diff
@@ -144,12 +144,19 @@ class DefinitionManager {
...
-        //     definition.settings.support_bottom_distance.default_value = supportZDistance;
-        // }
+        if (settings.support_z_distance) {
+            const supportZDistance = settings.support_z_distance.default_value;
+            definition.settings.support_top_distance.default_value = supportZDistance;
+            definition.settings.support_bottom_distance.default_value = supportZDistance / 2;
+        }
 
+        // fix CuraEngine z_overide_xy not effected on support_mesh
+        if (modelGroup && modelGroup.hasSupportModel() && settings.support_z_distance) {
```
Comment (parachvte) [inline src/app/flux/printing/DefinitionManager.js:154]: Only inject `hasSupportModel()` value, not `ModelGroup`.

File: src/app/flux/printing/index.js [hunk 2]
```diff
@@ -685,9 +688,9 @@ export const actions = {
             modelFileName: filename
         };
         */
+
         const params = {
-            originalName: originalName,
-            uploadName: uploadName,
+            ...result,
```
Comment (parachvte) [inline src/app/flux/printing/index.js:693]: Bad code changes, parameters should be explicitly defined.

File: src/app/flux/printing/index.js [hunk 3]
```diff
@@ -698,27 +701,41 @@ export const actions = {
...
+                    const formData = new FormData();
+                    formData.append('file', fileOfBlob);
+                    formData.append('uploadName', stlFileName);
+                    const uploadResult = await api.uploadFile(formData);
+                    if (item.supportTag === true) {
+                        ret.support.push(uploadResult.body.uploadName);
+                    } else {
+                        ret.model.push(uploadResult.body.uploadName);
+                        if (!ret.originalName) {
+                            ret.originalName = uploadResult.body.originalName;
```
Comment (parachvte) [inline src/app/flux/printing/index.js:733]: originalName not defined

File: src/app/flux/printing/index.js [hunk 4]
```diff
@@ -1069,6 +1094,11 @@ export const actions = {
                     const modelPositionAttribute = new THREE.BufferAttribute(positions, 3);
                     const material = new THREE.MeshPhongMaterial({ color: 0xa0a0a0, specular: 0xb0b0b0, shininess: 0 });
 
+                    // const material = new THREE.MeshPhongMaterial({
```
Comment (parachvte) [inline src/app/flux/printing/index.js:1097]: Remove if not being used

File: src/app/flux/printing/index.js [hunk 5]
```diff
@@ -698,27 +701,41 @@ export const actions = {
...
-                const uploadResult = await api.uploadFile(formData);
+                const models = modelGroup.models.filter(i => i.visible);
+                const ret = { model: [], support: [] };
+                for (const item of models) {
+                    const mesh = item.meshObject.clone();
+                    mesh.applyMatrix(item.meshObject.parent.matrix);
+                    const stl = new ModelExporter().parse(mesh, 'stl', true);
+                    const blob = new Blob([stl], { type: 'text/plain' });
+
+                    const originalName = item.originalName || (`supporter_${(Math.random() * 1000).toFixed(0)}`);
```
Comment (parachvte) [inline src/app/flux/printing/index.js:718]: Can we create originalName for support model on creation, rather do it here.

File: src/app/flux/printing/DefinitionManager.js [hunk 6]
```diff
@@ -144,12 +144,19 @@ class DefinitionManager {
...
         }
-        // if (settings.support_z_distance) {
-        //     const supportZDistance = settings.support_z_distance.default_value;
-        //     definition.settings.support_top_distance.default_value = supportZDistance;
-        //     definition.settings.support_bottom_distance.default_value = supportZDistance;
-        // }
+        if (settings.support_z_distance) {
+            const supportZDistance = settings.support_z_distance.default_value;
+            definition.settings.support_top_distance.default_value = supportZDistance;
+            definition.settings.support_bottom_distance.default_value = supportZDistance / 2;
```
Comment (parachvte) [inline src/app/flux/printing/DefinitionManager.js:150]: Why / 2?

File: src/app/models/Model.js [hunk 7]
```diff
@@ -460,14 +456,14 @@ class Model {
             // clone this.convexGeometry then clone.computeBoundingBox() is faster.
             if (this.convexGeometry) {
                 const clone = this.convexGeometry.clone();
-                this.meshObject.updateMatrix();
-                clone.applyMatrix(this.meshObject.matrix);
+                this.meshObject.updateMatrixWorld();
+                clone.applyMatrix(this.meshObject.matrixWorld);
```
Comment (parachvte) [inline src/app/models/Model.js:460]: Doubt

File: src/server/slicer/slice.js [hunk 8]
```diff
@@ -32,10 +32,28 @@ let curaEnginePath;
     }
 })();
 
-function callCuraEngine(modelPath, configPath, outputPath) {
+// extends api for support models
```
Comment (parachvte) [inline src/server/slicer/slice.js:35]: Add docs about the arguments

File: src/app/widgets/PrintingVisualizer/VisualizerModelTransformation.jsx [hunk 9]
```diff
@@ -29,13 +29,22 @@ class VisualizerModelTransformation extends PureComponent {
             scaleY: PropTypes.number,
             scaleZ: PropTypes.number
         }).isRequired,
+        defaultSupportSize: PropTypes.shape({
+            x: PropTypes.number,
+            y: PropTypes.number
+        }).isRequired,
+        isSupporting: PropTypes.bool.isRequired,
```
Comment (parachvte) [inline src/app/widgets/PrintingVisualizer/VisualizerModelTransformation.jsx:36]: Is supporting?

File: src/app/models/ModelGroup.js [hunk 10]
```diff
@@ -1,20 +1,19 @@
-// import { Euler, Vector3, Box3, Object3D } from 'three';
-// import { Vector3, Group, MeshPhongMaterial } from 'three';
-import { Vector3, Group, MeshPhongMaterial, Matrix4, Euler, Box3, Quaternion } from 'three';
+import * as THREE from 'three';
```
Comment (parachvte) [inline src/app/models/ModelGroup.js:1]: Revert the change

File: src/app/models/ModelGroup.js [hunk 11]
```diff
@@ -430,15 +450,21 @@ class ModelGroup extends EventEmitter {
         for (const model of this.models) {
             model.meshObject.removeEventListener('update', this.onModelUpdate);
             this.object.remove(model.meshObject);
+            // TODO: remove selected group without applyMatrix will cause problem
             this.selectedGroup.remove(model.meshObject);
         }
         this.models.splice(0);
         for (const model of models) {
-            const newModel = model.clone(this);
+            // remove clone model here
```
Comment (parachvte) [inline src/app/models/ModelGroup.js:458]: Bad comment

File: src/app/models/ModelGroup.js [hunk 12]
```diff
@@ -1090,6 +1142,9 @@ class ModelGroup extends EventEmitter {
         let isOverstepped = false;
         model.computeBoundingBox();
         isOverstepped = this._bbox && !this._bbox.containsBox(model.boundingBox);
+        if (isOverstepped) {
+            console.log(this._bbox, model.boundingBox);
```
Comment (parachvte) [inline src/app/models/ModelGroup.js:1146]: Remove log

File: src/app/components/three-extensions/ThreeUtils.js [hunk 13]
```diff
@@ -149,6 +149,39 @@ const ThreeUtils = {
...
         }
+    },
+
+    removeObjectParent(obj) {
+        const parent = obj.parent;
+        if(!parent) return;
+
+        parent.updateMatrixWorld();
+        parent.remove(obj);
+        obj.applyMatrix(parent.matrixWorld)
```
Comment (parachvte) [inline src/app/components/three-extensions/ThreeUtils.js:160]: Comma required

File: src/app/components/SMCanvas/SupportControls.js [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+class SupportControls extends THREE.Object3D {
+    camera = null;
+
+    raycaster = new THREE.Raycaster();
+
+    horizontalPlane = new THREE.Plane(new THREE.Vector3(0, 0, 1), 0);
+
+    _model = null;
+
+    constructor(camera, modelGroup) {
```
Comment (parachvte) [inline src/app/components/SMCanvas/SupportControls.js:15]: Not a good idea to inject model group!

File: src/app/components/SMCanvas/Controls.jsx [hunk 15]
```diff
@@ -250,13 +271,17 @@ class Controls extends EventEmitter {
         if (this.state !== STATE.NONE) {
             // Track events even when the mouse move outside of window
             document.addEventListener('mousemove', this.onDocumentMouseMove, false);
-            document.addEventListener('mouseup', this.onDocumentMouseUp, false);
+            // document.addEventListener('mouseup', this.onDocumentMouseUp, false);
```
Comment (parachvte) [inline src/app/components/SMCanvas/Controls.jsx:274]: Why

File: src/app/components/SMCanvas/Controls.jsx [hunk 16]
```diff
@@ -138,6 +148,7 @@ class Controls extends EventEmitter {
         this.domElement.addEventListener('mousemove', this.onMouseHover, false);
         this.domElement.addEventListener('wheel', this.onMouseWheel, false);
         this.domElement.addEventListener('click', this.onClick, false);
+        this.domElement.addEventListener('mouseup', this.onDocumentMouseUp, false);
```
Comment (parachvte) [inline src/app/components/SMCanvas/Controls.jsx:151]: Mouseup should be bound to document

File: src/app/components/SMCanvas/Controls.jsx [hunk 17]
```diff
@@ -91,7 +93,7 @@ class Controls extends EventEmitter {
     // Track if mouse moved during "mousedown" to "mouseup".
     mouseDownPosition = null;
 
-    constructor(sourceType, camera, group, domElement, onScale, onPan) {
+    constructor(sourceType, camera, group, domElement, onScale, onPan, modelGroup) {
```
Comment (parachvte) [inline src/app/components/SMCanvas/Controls.jsx:96]: Bad idea to inject model group.

File: src/app/flux/printing/DefinitionManager.js (hunk 6 above)
Comment (meiliang1123) [reply src/app/flux/printing/DefinitionManager.js:150]: cura does this, maybe better pringting result

File: src/app/models/Model.js (hunk 7 above)
Comment (meiliang1123) [reply src/app/models/Model.js:460]: It'll be alright~

File: src/app/widgets/PrintingVisualizer/VisualizerModelTransformation.jsx (hunk 9 above)
Comment (meiliang1123) [reply src/app/widgets/PrintingVisualizer/VisualizerModelTransformation.jsx:36]: yes, we need state of is-supporting to tell widget the canvas is doing support

File: src/app/components/SMCanvas/Controls.jsx (hunk 16 above)
Comment (meiliang1123) [reply src/app/components/SMCanvas/Controls.jsx:151]: reverted

File: src/app/components/SMCanvas/Controls.jsx (hunk 15 above)
Comment (meiliang1123) [reply src/app/components/SMCanvas/Controls.jsx:274]: reverted
